pkg/datapath/loader/netlink.go

```diff
@@ -444,3 +446,127 @@ func addHostDeviceAddr(hostDev netlink.Link, ipv4, ipv6 net.IP) error {
 	}
 	return nil
 }
+
+// setupIPIPDevices ensures the specified v4 and/or v6 devices are created and
+// configured with their respective sysctls.
+//
+// Calling this function may result in tunl0 (v4) or ip6tnl0 (v6) fallback
+// interfaces being created as a result of loading the ipip and ip6_tunnel
+// kernel modules by creating cilium_ tunnel interfaces. These are catch-all
+// interfaces for the ipip decapsulation stack. By default, these interfaces
+// will be created in new network namespaces, but Cilium disables this behaviour
+// by setting net.core.fb_tunnels_only_for_init_net = 2.
+//
+// In versions of Cilium prior to 1.15, the behaviour was as follows:
+//   - Repurpose the default tunl0 by setting it into collect_md mode and renaming
+//     it to cilium_ipip4. Use the interface for production traffic.
+//   - The same cannot be done for ip6tunl0, as collect_md cannot be enabled on
+//     this interface. Leave it unused.
+//   - Rename sit0 to cilium_sit, if present. This was potentially a mistake,
+//     as the sit module is not involved with ip6tnl interfaces.
+//
+// As of Cilium 1.15, if present, tunl0 is renamed to cilium_tunl and ip6tnl0 is
+// renamed to cilium_ip6tnl. Both devices are left unused. Only devices that were
+// explicitly created are used.
+func setupIPIPDevices(ipv4, ipv6 bool) error {
+	// FlowBased sets IFLA_IPTUN_COLLECT_METADATA, the equivalent of 'ip link add
+	// ... type ipip/ip6tnl external'. This is needed so bpf programs can use
+	// bpf_skb_[gs]et_tunnel_key() on packets flowing through tunnels.
+
+	if ipv4 {
+		// Set up IPv4 tunnel device if requested.
+		if err := setupDevice(&netlink.Iptun{
+			LinkAttrs: netlink.LinkAttrs{Name: defaults.CiliumIPIPv4Device},
+			FlowBased: true,
+		}); err != nil {
+			return fmt.Errorf("creating %s: %w", defaults.CiliumIPIPv4Device, err)
+		}
+
+		// Rename fallback device created by potential kernel module load after
+		// creating tunnel interface.
+		if err := renameDevice("tunl0", "cilium_tunl"); err != nil {
```

Comment:
Meta: we should stop relying on device names so much and instead track indices. That saves us from renames breaking stuff.